### PR TITLE
[Devtools] add logs for profiler tab switch & settings change

### PR DESCRIPTION
**Pull request by mondaychen**
Tested on a local build

This also contains a fix for the case when `missedEvent` contains metadata

### CODE REVIEW

File: packages/react-devtools-shared/src/devtools/views/Settings/SettingsContext.js [hunk 1]
```diff
@@ -68,6 +68,15 @@ type Context = {|
 const SettingsContext = createContext<Context>(((null: any): Context));
 SettingsContext.displayName = 'SettingsContext';
 
+function useLocalStorageWithLog<T>(
+  key: string,
+  initialValue: T | (() => T),
+): [T, (value: T | (() => T)) => void] {
+  return useLocalStorage<T>(key, initialValue, {
```
Comment (lunaruan) [inline packages/react-devtools-shared/src/devtools/views/Settings/SettingsContext.js:75]: nit: you probably can inline this:

```js
const [theme, setTheme] = useLocalStorage<Theme>('React::DevTools::theme', 'auto', SETTINGS_CHANGED_EVENT)

function useLocalStorage(key, initialValue, eventName?) {
   const loggerEvent = {
        event_name: eventName,
       ...
   }
}
```

File: packages/react-devtools-shared/src/Logger.js [hunk 2]
```diff
@@ -40,14 +40,20 @@ export type LogEvent =
     |}
   | {|
       +event_name: 'profiling-start',
+    |}
+  | {|
+      +event_name: 'profiler-tab-changed',
+    |}
+  | {|
+      +event_name: 'settings-changed',
```
Comment (lunaruan) [inline packages/react-devtools-shared/src/Logger.js:48]: Do we also want to put the other logs in here?

ex.

```
{|
    +event_name: 'settings-changed',
    +key: string,
    ...
|}
```

Comment (mondaychen) [reply packages/react-devtools-shared/src/Logger.js:48]: it's already in metadata :)

File: packages/react-devtools-shared/src/devtools/views/Settings/SettingsContext.js (hunk 1 above)
Comment (mondaychen) [reply packages/react-devtools-shared/src/devtools/views/Settings/SettingsContext.js:75]: But what if we want to log something other than the event name?

Comment (lunaruan) [reply packages/react-devtools-shared/src/devtools/views/Settings/SettingsContext.js:75]: What's an example? I'm not really that opinionated. Also down to pass in a loggerEvent and then append to it :)

File: packages/react-devtools-shared/src/Logger.js (hunk 2 above)
Comment (lunaruan) [reply packages/react-devtools-shared/src/Logger.js:48]: Hm, I think we should type the metadata too if you'd like to put it in metaData instead. Perhaps

```
{|
  +event_name: 'settings-changed',
  +metadata: {
     +key: string,
      ...
   }
|}
```

Comment (mondaychen) [reply packages/react-devtools-shared/src/Logger.js:48]: Metadata is currently a separate thing from the event object.
In my original design metadata is a freeform object that can be used or not used.
But yeah, I agree from this case we can tell it's better if they are defined together so we know for sure some events always have certain metadata attributes. I'll update it.

Comment (mondaychen) [reply packages/react-devtools-shared/src/Logger.js:57]: Updated :)

**Pull request merged**